sql/src/main/java/io/crate/analyze/relations/RelationAnalyzer.java

```diff
@@ -137,13 +139,79 @@ public AnalyzedRelation analyze(Node node, Analysis analysis) {
+            if (childRelation instanceof UnionSelect) {
+                qualifiedName = ((UnionSelect) childRelation).left().getQualifiedName().toString();
+            }
+            relationAnalysisContext.addSourceRelation(qualifiedName, childRelation);
+            ExpressionAnalyzer expressionAnalyzer = new ExpressionAnalyzer(
+                functions,
+                statementContext.transactionContext(),
+                statementContext.convertParamFunction(),
+                new FullQualifiedNameFieldProvider(
+                    relationAnalysisContext.sources(),
+                    relationAnalysisContext.parentSources(),
+                    statementContext.transactionContext().sessionContext().defaultSchema()),
+                new SubqueryAnalyzer(this, statementContext));
+            ExpressionAnalysisContext expressionAnalysisContext = relationAnalysisContext.expressionAnalysisContext();
+            SelectAnalysis selectAnalysis = new SelectAnalysis(
+                childRelation.fields().size(),
+                relationAnalysisContext.sources(),
+                expressionAnalyzer,
+                expressionAnalysisContext);
+            for (Field field : childRelation.fields()) {
+                selectAnalysis.add(field.path(), field);
+            }
+            statementContext.endRelation();
+
+            return new OrderedLimitedRelation(
+                    childRelation,
+                    analyzeOrderBy(
+                        selectAnalysis,
+                        node.getOrderBy(),
+                        expressionAnalyzer,
+                        expressionAnalysisContext,
+                        false,
+                        false),
+                    longSymbolOrNull(node.getLimit(), expressionAnalyzer, expressionAnalysisContext),
+                    longSymbolOrNull(node.getOffset(), expressionAnalyzer, expressionAnalysisContext));
+        }
+        return process(node.getQueryBody(), statementContext);
     }
 
     @Override
     protected AnalyzedRelation visitUnion(Union node, StatementAnalysisContext context) {
-        throw new UnsupportedFeatureException("UNION is not supported");
+        if (node.isDistinct()) {
+            throw new UnsupportedFeatureException("UNION [DISTINCT] is not supported");
+        }
+        QueriedRelation first = (QueriedRelation) process(node.getLeft(), context);
+        QueriedRelation second = (QueriedRelation) process(node.getRight(), context);
+
+        // Check number of outputs
+        if (first.querySpec().outputs().size() != second.querySpec().outputs().size()) {
+            throw new UnsupportedOperationException("Number of output columns must be the same for all parts of a UNION");
+        }
+        // Try to cast outputs
+        List<DataType> dataTypesFromLeft = new ArrayList<>();
+        for (Symbol outputSymbol : first.querySpec().outputs()) {
```

Comment:
Could use the `typeView` function to avoid the copy but just create a view.
Might even want to move this code into a `ensureOutputsHaveCompatibleTypes` method or something - would make it clear enough that the comment isn't required.